autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/importing/PasswordImporter.kt

```diff
@@ -0,0 +1,122 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.autofill.impl.importing
+
+import android.os.Parcelable
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.autofill.api.domain.app.LoginCredentials
+import com.duckduckgo.autofill.impl.importing.PasswordImporter.ImportResult
+import com.duckduckgo.autofill.impl.importing.PasswordImporter.ImportResult.Finished
+import com.duckduckgo.autofill.impl.importing.PasswordImporter.ImportResult.InProgress
+import com.duckduckgo.autofill.impl.store.InternalAutofillStore
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.di.scopes.AppScope
+import com.squareup.anvil.annotations.ContributesBinding
+import dagger.SingleInstanceIn
+import java.util.UUID
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.parcelize.Parcelize
+
+interface PasswordImporter {
+    suspend fun importPasswords(importList: List<LoginCredentials>, originalImportListSize: Int): String
+    fun getImportStatus(jobId: String): Flow<ImportResult>
+
+    sealed interface ImportResult : Parcelable {
+        val jobId: String
+
+        @Parcelize
+        data class InProgress(
+            val savedCredentialIds: List<Long>,
+            val numberSkipped: Int,
+            val importListOriginalSize: Int,
+            override val jobId: String,
+        ) : ImportResult
+
+        @Parcelize
+        data class Finished(
+            val savedCredentialIds: List<Long>,
+            val numberSkipped: Int,
+            val importListOriginalSize: Int,
+            override val jobId: String,
+        ) : ImportResult
+    }
+}
+
+@SingleInstanceIn(AppScope::class)
+@ContributesBinding(AppScope::class)
+class PasswordImporterImpl @Inject constructor(
+    private val existingPasswordMatchDetector: ExistingPasswordMatchDetector,
+    private val autofillStore: InternalAutofillStore,
+    private val dispatchers: DispatcherProvider,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+) : PasswordImporter {
+
+    private val _importStatus = MutableSharedFlow<ImportResult>(replay = 1)
+    private val mutex = Mutex()
+
+    override suspend fun importPasswords(importList: List<LoginCredentials>, originalImportListSize: Int): String {
+        val jobId = UUID.randomUUID().toString()
+
+        mutex.withLock {
```

Comment:
I believe the intention here is to control we don't import while we are importing...right? If so the implementation is not correct because `launch` is fire and forget, and will automatically release the lock. That means in practice we execute as many imports as we trigger.

I think the logic was just for something specific on the internal screens, so I'm not sure if it will be still present in following PRs. Just questioning if we still need this kind of logic here in production. What is the scenario we were trying to solve excluding the internal case?

Reply:
Changed this to be simpler